stackslib/src/net/api/blocksimulate.rs

```diff
@@ -0,0 +1,380 @@
+    pub tx_merkle_root: Sha512Trunc256Sum,
+    pub state_index_root: TrieHash,
+    pub timestamp: u64,
+    pub miner_signature: MessageSignature,
+    pub signer_signature: Vec<MessageSignature>,
+    pub transactions: Vec<RPCSimulatedBlockTransaction>,
+    pub valid_merkle_root: bool,
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCNakamotoBlockSimulateRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(r#"^/v3/blocks/simulate/(?P<block_id>[0-9a-f]{64})$"#).unwrap()
+    }
+
+    fn metrics_identifier(&self) -> &str {
+        "/v3/blocks/simulate/:block_id"
+    }
+
+    /// Try to decode this request.
+    /// There's nothing to load here, so just make sure the request is well-formed.
+    fn try_parse_request(
+        &mut self,
+        preamble: &HttpRequestPreamble,
+        captures: &Captures,
+        query: Option<&str>,
+        _body: &[u8],
+    ) -> Result<HttpRequestContents, Error> {
+        // If no authorization is set, then the block simulation endpoint is not enabled
+        let Some(password) = &self.auth else {
+            return Err(Error::Http(400, "Bad Request.".into()));
+        };
+        let Some(auth_header) = preamble.headers.get("authorization") else {
+            return Err(Error::Http(401, "Unauthorized".into()));
+        };
+        if auth_header != password {
+            return Err(Error::Http(401, "Unauthorized".into()));
+        }
+        if preamble.get_content_length() != 0 {
+            return Err(Error::DecodeError(
+                "Invalid Http request: expected 0-length body".to_string(),
+            ));
+        }
+
+        let block_id_str = captures
+            .name("block_id")
+            .ok_or_else(|| {
+                Error::DecodeError("Failed to match path to block ID group".to_string())
+            })?
+            .as_str();
+
+        let block_id = StacksBlockId::from_hex(block_id_str).map_err(|_| {
+            Error::DecodeError("Invalid path: unparseable consensus hash".to_string())
+        })?;
+        self.block_id = Some(block_id);
+
+        Ok(HttpRequestContents::new().query_string(query))
+    }
+}
+
+impl RPCRequestHandler for RPCNakamotoBlockSimulateRequestHandler {
+    /// Reset internal state
+    fn restart(&mut self) {
+        self.block_id = None;
+    }
+
+    /// Make the response
+    fn try_handle_request(
+        &mut self,
+        preamble: HttpRequestPreamble,
+        _contents: HttpRequestContents,
+        node: &mut StacksNodeState,
+    ) -> Result<(HttpResponsePreamble, HttpResponseContents), NetError> {
+        let block_id = self
+            .block_id
+            .take()
+            .ok_or(NetError::SendError("Missing `block_id`".into()))?;
+
+        let simulated_block_res =
+            node.with_node_state(|_network, sortdb, chainstate, _mempool, _rpc_args| {
+                let block_id = block_id.clone();
```

Comment:
In order to better test this (and factor it), can you put the code body of this closure into its own function?

Reply:
For example, you could verify that you get the errors leading to HTTP 404 or HTTP 500 if you factored this into its own function, and then added more thorough unit tests for this that don't (necessarily) rely on `TestRPC`.